benchmark/_flocker.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2016 ClusterHQ Inc.  See LICENSE file for details.
+"""
+Utilities to perform Flocker operations.
+"""
+
+from functools import partial
+from datetime import timedelta
+
+from flocker.common import loop_until, timeout as _timeout
+
+DEFAULT_TIMEOUT = timedelta(minutes=10)
+
+
+def loop_until_state_found(reactor, get_states, state_matches, timeout):
+    """
+    Loop until a state has been reached.
+
+    :param get_states: Callable returning a Deferred firing with a list
+        of states.
+    :param state_matches: Callable that accepts a state parameter, and
+        returns a boolean indicating whether the state matches.
+    :param timedelta timeout: Maximum time to wait for state to be found.
+    :return Deferred[Any]: The matching state.
+    """
+    def state_reached():
+        d = get_states()
+
+        def find_match(states):
+            for state in states:
+                if state_matches(state):
+                    return state
+            return None
+        d.addCallback(find_match)
+
+        return d
+
+    d = loop_until(reactor, state_reached)
+    _timeout(reactor, d, timeout.total_seconds())
```

Comment:
Thanks. Also would have been useful to call it `deferred_timeout` or similar to avoid the common name clash with `timeout` for the value.

Maybe we could add a timedelta-based version with a longer name.